src/pkcs15init/pkcs15-openpgp.c

```diff
@@ -185,46 +201,167 @@ static int openpgp_generate_key(sc_profile_t *profile, sc_pkcs15_card_t *p15card
 		return SC_ERROR_NOT_SUPPORTED;
 	}
 
-	/* Prepare buffer */
+
 	key_info.algorithm = SC_OPENPGP_KEYALGO_RSA;
-	key_info.rsa.modulus_len = required->modulus_length;
-	key_info.rsa.modulus = calloc(required->modulus_length >> 3, 1);
-	if (key_info.rsa.modulus == NULL)
+
+	/* Prepare buffer */
+	key_info.u.rsa.modulus_len = required->modulus_length;
+	key_info.u.rsa.modulus = calloc(required->modulus_length >> 3, 1);
+	if (key_info.u.rsa.modulus == NULL)
 		LOG_FUNC_RETURN(ctx, SC_ERROR_NOT_ENOUGH_MEMORY);
 
 	/* The OpenPGP supports only 32-bit exponent. */
-	key_info.rsa.exponent_len = 32;
-	key_info.rsa.exponent = calloc(key_info.rsa.exponent_len>>3, 1); /* 1/8 */
-	if (key_info.rsa.exponent == NULL) {
-		free(key_info.rsa.modulus);
+	key_info.u.rsa.exponent_len = 32;
+	key_info.u.rsa.exponent = calloc(key_info.u.rsa.exponent_len>>3, 1); /* 1/8 */
+	if (key_info.u.rsa.exponent == NULL) {
+		free(key_info.u.rsa.modulus);
 		LOG_FUNC_RETURN(ctx, SC_ERROR_NOT_ENOUGH_MEMORY);
 	}
 
 	r = sc_card_ctl(card, SC_CARDCTL_OPENPGP_GENERATE_KEY, &key_info);
-	if (r < 0)
-		goto out;
+	LOG_TEST_GOTO_ERR(card->ctx, r, "on-card EC key generation failed");
 
 	pubkey->algorithm = SC_ALGORITHM_RSA;
-
 	sc_log(ctx, "Set output modulus info");
-	pubkey->u.rsa.modulus.len = key_info.rsa.modulus_len;
-	pubkey->u.rsa.modulus.data = calloc(key_info.rsa.modulus_len, 1);
+	pubkey->u.rsa.modulus.len = key_info.u.rsa.modulus_len;
+	pubkey->u.rsa.modulus.data = calloc(key_info.u.rsa.modulus_len, 1);
 	if (pubkey->u.rsa.modulus.data == NULL)
-		goto out;
-	memcpy(pubkey->u.rsa.modulus.data, key_info.rsa.modulus, key_info.rsa.modulus_len);
+		goto err;
+	memcpy(pubkey->u.rsa.modulus.data, key_info.u.rsa.modulus, key_info.u.rsa.modulus_len);
 
 	sc_log(ctx, "Set output exponent info");
-	pubkey->u.rsa.exponent.len = key_info.rsa.exponent_len;
-	pubkey->u.rsa.exponent.data = calloc(key_info.rsa.exponent_len>>3, 1); /* 1/8 */
+	pubkey->u.rsa.exponent.len = key_info.u.rsa.exponent_len;
+	pubkey->u.rsa.exponent.data = calloc(key_info.u.rsa.exponent_len>>3, 1); /* 1/8 */
 	if (pubkey->u.rsa.exponent.data == NULL)
-		goto out;
-	memcpy(pubkey->u.rsa.exponent.data, key_info.rsa.exponent, key_info.rsa.exponent_len>>3); /* 1/8 */
-
-out:
-	if (key_info.rsa.modulus)
-		free(key_info.rsa.modulus);
-	if (key_info.rsa.exponent)
-		free(key_info.rsa.exponent);
+		goto err;
+	memcpy(pubkey->u.rsa.exponent.data, key_info.u.rsa.exponent, key_info.u.rsa.exponent_len>>3); /* 1/8 */
```

Comment:
What happens if the length is not a multiple of 8?

Reply:
What do you suggest? I don't like this bit pushing either, but I don't know how to prevent it (and just copied it from previous code). 
Do you like me to just test `len%8 == 0` each time? Or do you have an idea to fix this issue generally?

Reply:
@alex-nitrokey To make sure there are enough bytes available even for bit lengths which are not a multiple of 8, simply use the formula: `number_of_bytes = (number_of_bits + 7) / 8`.

Maybe a macro can help to increase code legibility, e.g.:
```
#define BYTES4BITS(num)  (((num) + 7) / 8)    /* number of bytes necessary to hold 'num' bits */
```

Reply:
Fixed via e706b3ba
I was not sure, what the best location for the define is and I put it in internal.h.
Do you have better idea?